velox/connectors/kafka/KafkaDataSource.cpp

```diff
@@ -0,0 +1,151 @@
+            createConsumer(cppKafkaConfig);
+        }
+        createMessageQueue(consumeBatchSize_);
+        createRecordDeserializer(config_->getFormat(), outputType_);
+    }
+
+    bool KafkaDataSource::consumerCanbeCreated() {
+        return config_->exists(ConnectionConfig::kBootstrapServers)
+            && config_->exists(ConnectionConfig::kClientId)
+            && config_->exists(ConnectionConfig::kTopic)
+            && config_->exists(ConnectionConfig::kGroupId)
+            && config_->exists(ConnectionConfig::kFormat)
+            && !consumer_.get();
+    }
+
+    void KafkaDataSource::createConsumer(cppkafka::Configuration & config) {
+        VELOX_CHECK_NULL(consumer_.get(), "Failed to create kafka consumer as the consumer is not null");
+        CppKafkaConsumerPtr cppKafkaConsumer = std::make_shared<cppkafka::Consumer>(config);
+        cppKafkaConsumer->set_destroy_flags(RD_KAFKA_DESTROY_F_NO_CONSUMER_CLOSE);
+        consumer_ = std::make_shared<KafkaConsumer>(cppKafkaConsumer, config_->getPollTimeoutMills(), consumeBatchSize_);
+        String topic = config_->getTopic();
+        topics_.emplace_back(topic);
+        consumer_->subscribe(topics_);
+    }
+
+    void KafkaDataSource::createMessageQueue(const uint32_t size) {
+        VELOX_CHECK_GT(size, 0, "Kafka consume message queue size must greater than 0");
+        queue_.reserve(size);
+    }
+
+    void KafkaDataSource::createRecordDeserializer(const String & format, const RowTypePtr & outputType) {
+        if (format == "json") {
+            deserializer_ = std::make_shared<KafkaJSONRecordDeserializer>(outputType, queryCtx_->memoryPool()); 
+        } else if (format == "csv") {
+            deserializer_ = std::make_shared<KafkaCSVRecordDeserializer>(outputType, queryCtx_->memoryPool());
+        } else if (format == "raw") {
+            deserializer_ = std::make_shared<KafkaRawRecordDeserializer>(outputType, queryCtx_->memoryPool());
+        } else {
+            VELOX_FAIL_UNSUPPORTED_INPUT_UNCATCHABLE("The data format {} is not supported for kafka.", format);
+        }
+        emptyRow_ = deserializer_->emptyRow();
+        outRow_ = deserializer_->emptyRow();
+        outRow_->resize(1);
+    }
+
+    void KafkaDataSource::addSplit(ConnectorSplitPtr split) {
+        KafkaConnectorSplit * kafkaConnectorSplit = static_cast<KafkaConnectorSplit *>(split.get());
+        VELOX_CHECK_NOT_NULL(kafkaConnectorSplit, "Failed to add split, because the kafka connector split is null.");
+        VELOX_CHECK_NOT_NULL(consumer_.get(), "Failed to add split, because the kafka consumer is null.");
+        cppkafka::TopicPartitionList topicPartitions = kafkaConnectorSplit->getCppKafkaTopicPartitions();
+        consumer_->assign(topicPartitions);
+    }
+
+    std::optional<RowVectorPtr> KafkaDataSource::next(uint64_t, velox::ContinueFuture&) {
+        std::optional<RowVectorPtr> res;
+        if (processByBatch_) {
```

Comment:
What is processByBatch_ meaning? You just need to put the consumed messages in a RowVector.